### PR TITLE
Public registration & oAuth2 \o/

### PR DESCRIPTION
**Pull request by nicosomb**
* public registration
* remove WSSE implementation
* add oAuth2 implementation

### CODE REVIEW

**Comment by nicosomb** [comment]
Sorry, I created a big PR.

**Comment by nicosomb** [comment]
Todo: 
- [ ] a nice form for registration
- [ ] we can't access to /api/doc and I don't know why
- [ ] tests

**Comment by j0k3r** [comment]
Wow, good job.

**Comment by nicosomb** [comment]
I need to talk with you about passwords.

File: src/Wallabag/CoreBundle/EventListener/AuthenticationListener.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        );
+    }
+
+    public function authenticate(FilterUserResponseEvent $event, $eventName = null, EventDispatcherInterface $eventDispatcher = null)
+    {
+        if (!$event->getUser()->isEnabled()) {
+            return;
+        }
+
+        $config = new Config($event->getUser());
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/EventListener/AuthenticationListener.php]: Could you put a comment on that class to explain that it is called after a user is registered?
Or maybe renamed it to `RegistratioConfirmedListener` to be more explicit.

Comment (nicosomb) [inline src/Wallabag/CoreBundle/EventListener/AuthenticationListener.php]: Renamed.

**Comment by nicosomb** [comment]
I added tests for public registration. Do you think it's enough for this feature? And for this PR?

**Comment by j0k3r** [comment]
You can add unit test for `RegistrationConfirmedListener`.
And you can run `phpunit --coverage-html build/coverage`, wait 30 minutes (...) and see where you still need some tests

**Comment by nicosomb** [comment]
Added a test on `RegistrationConfirmedListener`.

**Comment by j0k3r** [comment]
Thanks for the _functional_ test but I was talking about a unit test (which doesn't load all Symfony) but only the `RegistrationConfirmedListener` class. Like in `UsernameRssTokenConverterTest`.

**Comment by nicosomb** [comment]
oops, sorry for the container :-(

**Comment by nicosomb** [comment]
and thank you for the test :+1: ! 
This PR is ready for review !

**Comment by j0k3r** [comment]
Next step: remove what we've done about forgot password, login, etc ..

**Comment by nicosomb** [comment]
Planned. And also create a UserBundle.

**Comment by nicosomb** [comment]
Ooooh I don't know what I did :( all these commits ...

**Comment by j0k3r** [comment]
I think I've fixed the mess :)
I guess you killed your repo by rebasing the branch against the v2, maybe in a wrong way.

I've cherry-picked your 2 commits and fixed few things.
Also, now with the UserBundle, testing with coverage is like ... taking ages ... :confused: 

```
PHPUnit 4.8.10 by Sebastian Bergmann and contributors.

................................................................. 65 / 86 ( 75%)
.....................

Time: 25.59 minutes, Memory: 287.50Mb

OK (86 tests, 337 assertions)

Generating code coverage report in HTML format ... done
```

**Comment by nicosomb** [comment]
thank you for your help!

**Comment by j0k3r** [comment]
There are still some work to be done :)

- [ ] graby registration page
- [ ] a better forgot password page
- [ ] a functional test on registration (because we use a, even small, custom form type)

**Comment by nicosomb** [comment]
> graby registration page

What is it? 

I will work on the other tasks soon.

**Comment by j0k3r** [comment]
Woops, I said graby but I wanted to say _baggy_ :smile:

**Comment by tcitworld** [comment]
Why is the name field a textarea ? Be sure to make it optional.

**Comment by nicosomb** [comment]
It's already deleted in next commits ;-)

**Comment by nicosomb** [comment]
> a functional test on registration (because we use a, even small, custom form type)

I removed `RegistrationType.php`. Is the test still necessary? 

I have to work on baggy form, but material is now OK :-)

**Comment by j0k3r** [comment]
I tried to remove `RegistrationType` but got an error in the test suite ..

**Comment by nicosomb** [comment]
RegistrationType was removed in my last commit and test suite seems to be green.

**Comment by nicosomb** [comment]
Ready for last review.

**Comment by j0k3r** [comment]
Should be ok for now, we'll see later if we need improvment.
Good job :+1:

**Pull request merged**